### ISSUE TITLE
How to deal with the day-to-day oscillation of the taxi dispatcher?

### ISSUE THREAD

**Issue by kowinter** [matsim-org/matsim-code-examples#148]
Dear all,
I have an issue with the taxi dispatcher: it produces very different results for the final day of a series of simulated days and for the first day of a new series, when taking the output-plans of the first series as input of the second series. As an illustration of the problem, I attach the "status time profiles" of the final day of round one (75th run) 
![75 taxi_status_time_profiles_StackedArea](https://user-images.githubusercontent.com/23402187/59430332-474f1080-8de2-11e9-83b3-600a5495b549.png) 
and the first day of the second round (0th run).
![0 taxi_status_time_profiles_StackedArea](https://user-images.githubusercontent.com/23402187/59430319-428a5c80-8de2-11e9-9d46-26f4c53058dc.png)

My goal is to take the output of the first series and use it as input for various scenarios, in order to shorten run-times. Due to computation time, I cannot afford to let all scenarios evolve from scratch.

As I learned from @jfbischoff , the day-to-day oscillation has to do with the dispatcher "learning" the best routing options over time, something that cannot be turned off, I assume. I see 2 options on how to deal with this problem: (1) providing the dispatcher in the second round with the information on routing used by the dispatcher in the first round or (2) simulate a series of days in the second round in which only the dispatcher "learns", while turning of any learning behavior of the agents. The second option is not really satisfying in my case, as it takes more than 25 simulated days (and more than 30h) to come even close to a solution that is in the range of the final day of the first series, defeating the purpose of splitting the simulation into 2 parts. 

Are there any examples out there on how to deal with this issue? Does somebody know how to access the information based on which the dispatcher "learns" how to do the routing, and how to feed this into the dispatcher in a new round of simulations? 

Any thoughts on this are very much appreciated!

**In short: The ultimate goal is to re-create in one simulated day the last day of a series of simulated days while using the taxi dispatcher.**

**Comment by michalmac** [comment]
The cause is that the link travel time information used by taxis in the last day of the 1st series is lost.

Link travel times used in DVRP are exponential moving averages over all iterations (with the last iteration having the highest weight). This prevents oscillations. For iter. 0, DVRP link travel times are initialised assuming free flow of vehicles.

There are two ways to cope with your issue:
1. Initialising DVRP link travel times with some reasonable values (in your case, these would be values used in the last day of the 1st series, which should give almost the same results -- there might be some variations due to random components).
2. Use of online travel time calculation, where link travel times are calculated both using the historical values (i.e. the exponential moving average) and the recently observed travel times (e.g. in the last quarter or an hour).

The former option requires implementing writing and reading DVRP travel times to/from a file.

The latter option is easier to start with - no extra programming needed. You only need to adjust the following 2 parameters in the DVRP config group: https://github.com/matsim-org/matsim/blob/43211ffc654463f568ef05948b439c3e396b1747/contribs/dvrp/src/main/java/org/matsim/contrib/dvrp/run/DvrpConfigGroup.java#L56-L87. However, this is not exact to:
> re-create in one simulated day the last day of a series of simulated days

**Comment by michalmac** [comment]
A small variation of (1):

You can read in events from the last iteration and compute the link travel times experienced in the last iteration (this will not be exactly to the DVRP TTs used in the last iteration, but should be close to them given the scenario was relaxed):
```
TravelTime experiencedTravelTimes = TravelTimeUtils.createTravelTimesFromEvents(scenario, eventsFile);
```
Then you need to bind the experienced travel times as the initial DVRP travel times:
```
controller.addOverridingModule(new AbstractModule() {
    public void install() {
            addTravelTimeBinding(DvrpTravelTimeModule.DVRP_INITIAL).toInstance(experiencedTravelTimes);
    }
}
```

**Comment by kowinter** [comment]
Dear Michal,

Thank you very much for your prompt reply! 
I tried both options (also in combination) for various scenarios - unfortunately it only improves the dispatcher marginally, and the issue of the miss-guided/stuck taxis remains. If I should come across a solution I will post this here. 

Best,
K.

**Comment by michalmac** [comment]
We used this simplified approach: https://github.com/matsim-org/matsim-code-examples/issues/148#issuecomment-501864910 for our paper on AT congestions (https://www.researchgate.net/publication/319489487_Congestion_effects_of_autonomous_taxi_fleets). The 0th iteration with ATs was started from a relaxed scenario. In the worst case (replacement of all car trips by AT trips => introduction of a lot of empty-vehicle traffic) we needed 10 iterations to reach a more/less stable traffic. In many cases, it was only a few iterations. But as I said, we used the simplified approach.

**Comment by michalmac** [comment]
Have you also tried this option?

>Initialising DVRP link travel times with some reasonable values (in your case, these would be values used in the last day of the 1st series, which should give almost the same results -- there might be some variations due to random components).

**Comment by kowinter** [comment]
Thanks for your response!


> We used this simplified approach: [#148 (comment)](https://github.com/matsim-org/matsim-code-examples/issues/148#issuecomment-501864910) for our paper on AT congestions (https://www.researchgate.net/publication/319489487_Congestion_effects_of_autonomous_taxi_fleets). The 0th iteration with ATs was started from a relaxed scenario. In the worst case (replacement of all car trips by AT trips => introduction of a lot of empty-vehicle traffic) we needed 10 iterations to reach a more/less stable traffic. In many cases, it was only a few iterations. But as I said, we used the simplified approach.

--> Unfortunately I need much more than 10 iterations if I turn the mode-switching behavior off (which I would like to do). If I keep the mode-switching behavior on for 10% of the population (remembering 3 plans), I can get stable results after approx. 7 iterations.



> Have you also tried this option?
> 
> > Initialising DVRP link travel times with some reasonable values (in your case, these would be values used in the last day of the 1st series, which should give almost the same results -- there might be some variations due to random components).

--> only in form of the variation you had mentioned above, as I don't have know how to get the exact values the dispatcher used on the last day of the 1st series. Would that be the travel times from the second-last day of the 1st series?

**Comment by kowinter** [comment]
**Update -issue solved in a 3 step process**

I solved the issue by performing the following steps:

(1) first round: let modal share for taxi evolve based on desired learning behavior for many days
(2) second round: 15 or 20 runs in which mode change behavior is still kept on, but 'fractionOfIterationsToDisableInnovation' is turned off (important!), with only the selected plans  (not the whole set of remembered plans) and the events of the last day (as described by @michalmac) of round 1 as input. In this round the plans won't change drastically over the low number of  runs - close enough is good enough...

> A small variation of (1):
> 
> You can read in events from the last iteration and compute the link travel times experienced in the last iteration (this will not be exactly to the DVRP TTs used in the last iteration, but should be close to them given the scenario was relaxed):
> 
> ```
> TravelTime experiencedTravelTimes = TravelTimeUtils.createTravelTimesFromEvents(scenario, eventsFile);
> ```
> 
> Then you need to bind the experienced travel times as the initial DVRP travel times:
> 
> ```
> controller.addOverridingModule(new AbstractModule() {
>     public void install() {
>             addTravelTimeBinding(DvrpTravelTimeModule.DVRP_INITIAL).toInstance(experiencedTravelTimes);
>     }
> }
> ```

(3) third round: simulate 1 day with the selected plans of the last day of round 2 as input 

Maybe the 3-step process could been shortened to a 2-step process if 'fractionOfIterationsToDisableInnovation' would have been turned off in the first round -worthwhile trying the next time.

Thanks again @jfbischoff  and @michalmac for having looked into this!
